community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/planner/logical/CardinalityCostModel.scala

```diff
@@ -19,47 +19,69 @@
  */
 package org.neo4j.cypher.internal.compiler.v2_2.planner.logical
 
-import org.neo4j.cypher.internal.compiler.v2_2.ast.Collection
+import org.neo4j.cypher.internal.compiler.v2_2.ast.{HasLabels, Property, Collection}
 import org.neo4j.cypher.internal.compiler.v2_2.commands.ManyQueryExpression
 import org.neo4j.cypher.internal.compiler.v2_2.planner.logical.Metrics._
 import org.neo4j.cypher.internal.compiler.v2_2.planner.logical.plans._
 
-/*
-A very simplistic cost model. Each row returned by an operator costs 1. That's it.
- */
 case class CardinalityCostModel(cardinality: CardinalityModel) extends CostModel {
 
-  val CPU_BOUND_PLAN_COST_PER_ROW: CostPerRow = 0.1
-  val DB_ACCESS_BOUND_PLAN_COST_PER_ROW: CostPerRow = 1.0
+  val CPU_BOUND:  CostPerRow = 0.1
+  val FAST_STORE: CostPerRow = 1.0
+  val SLOW_STORE: CostPerRow = 10.0
 
   private def costPerRow(plan: LogicalPlan): CostPerRow = plan match {
 
+    case _: AllNodesScan |
+         _: DirectedRelationshipByIdSeek |
+         _: UndirectedRelationshipByIdSeek |
+         _: ProjectEndpoints |
+         _: NodeByLabelScan
+    => FAST_STORE
+
+    // Filtering on labels and properties
+    case Selection(predicates, _) if predicates.exists {
+      case _:Property | _: HasLabels => true
+      case _ => false
+    }
+    => FAST_STORE
+
     case _: NodeHashJoin |
-      _: Aggregation |
-      _: AbstractLetSemiApply |
-      _: Limit |
-      _: Optional |
-      _: SingleRow |
-      _: Argument |
-      _: OuterHashJoin |
-      _: AbstractSemiApply |
-      _: Skip |
-      _: Sort |
-      _: SortedLimit |
-      _: Union |
-      _: UnwindCollection
-    => CPU_BOUND_PLAN_COST_PER_ROW
-
-    case NodeIndexSeek(_, _, _, ManyQueryExpression(Collection(elements)), _) =>
-      DB_ACCESS_BOUND_PLAN_COST_PER_ROW * Multiplier(elements.size)
+         _: Aggregation |
+         _: AbstractLetSemiApply |
+         _: Limit |
+         _: Optional |
+         _: SingleRow |
+         _: Argument |
+         _: OuterHashJoin |
+         _: AbstractSemiApply |
+         _: Skip |
+         _: Sort |
+         _: SortedLimit |
+         _: Union |
+         _: Selection |
+         _: UnwindCollection
+    => CPU_BOUND
 
-    case _ =>
-      DB_ACCESS_BOUND_PLAN_COST_PER_ROW
+    case _: Expand |
+         _: FindShortestPaths |
+         _: LegacyIndexSeek |
+         _: NodeByIdSeek |
+         _: NodeIndexUniqueSeek |
+         _: NodeIndexSeek |
+         _: VarExpand
+    => SLOW_STORE
```

Comment:
Some comments explaining why such stores are slow compared with other would be very nice for future reference.

E.g., I can see why index seeks are considered slow, but I am less sure about expands.